flink-libraries/flink-python/src/main/python/org/apache/flink/python/api/flink/plan/DataSet.py

```diff
@@ -572,6 +572,41 @@ def set_parallelism(self, parallelism):
         self._info.parallelism.value = parallelism
         return self
 
+    def count_elements_per_partition(self):
+        class CountElementsPerPartitionMapper(MapPartitionFunction):
+
+            def map_partition(self, iterator, collector):
+                counter = 0
+                for x in iterator:
+                    counter += 1
+
+                collector.collect((self.context.get_index_of_this_subtask(), counter))
+        return self.map_partition(CountElementsPerPartitionMapper())
+
+    def zip_with_index(self):
+        element_count = self.count_elements_per_partition()
+        class ZipWithIndexMapper(MapPartitionFunction):
+            start = 0
```

Comment:
if we would start at -1 we wouldn't have to decrement it when collecting the value